doc/sphinx/proof-engine/tactics.rst

```diff
@@ -1676,68 +1615,100 @@ Controlling the proof flow
+   :name: generalize; _
 
-   .. coqtop:: reset none
+   For a single :n:`@one_term`, replaces the goal `G` with `forall (x:T), G'`,
+   where :n:`@one_term` is a subterm of `G` of type `T` and `G'` is obtained
+   by replacing all occurrences of :n:`@one_term` with `x` within `G`.  `x` is
+   a fresh variable chosen based on `T`.  Specifying multiple :n:`one_term`\s is
+   equivalent to :n:`generalize @term__n; … ; generalize @term__1`.  (Note they
+   are processed *right to left*.)
 
-      Goal forall x y:nat, 0 <= x + y + y.
-      Proof. intros *.
+   :n:`@occs_nums`
+     Selects which occurrences of :n:`@one_term` to replace (counting from
+     left to right on the expression printed using the :flag:`Printing All` flag)
 
-   .. coqtop:: all abort
+   :n:`@as_name`
+     The name to use for `x` instead of a fresh name.
 
-      Show.
-      generalize (x + y + y).
+   :n:`@pattern_occs`
+     Limits the replacement to terms that match the :n:`@one_term` within :n:`@pattern_occs`.
 
-If the goal is :g:`G` and :g:`t` is a subterm of type :g:`T` in the goal,
-then :n:`generalize t` replaces the goal by :g:`forall (x:T), G′` where :g:`G′`
-is obtained from :g:`G` by replacing all occurrences of :g:`t` by :g:`x`. The
-name of the variable (here :g:`n`) is chosen based on :g:`T`.
+   .. example::
 
-.. tacv:: generalize {+ @term}
+      .. coqtop:: reset none
 
-   This is equivalent to :n:`generalize @term; ... ; generalize @term`.
-   Note that the sequence of term :sub:`i` 's are processed from n to 1.
+         Goal forall x y:nat, 0 <= x + y + y.
+         Proof. intros *.
 
-.. tacv:: generalize @term at {+ @natural}
+      .. coqtop:: all abort
 
-   This is equivalent to :n:`generalize @term` but it generalizes only over the
-   specified occurrences of :n:`@term` (counting from left to right on the
-   expression printed using the :flag:`Printing All` flag).
+         Show.
+         generalize (x + y + y).
 
-.. tacv:: generalize @term as @ident
+   .. tacn:: generalize dependent @one_term
 
-   This is equivalent to :n:`generalize @term` but it uses :n:`@ident` to name
-   the generalized hypothesis.
+      Generalizes :n:`@one_term` and all hypotheses that depend on :n:`@one_term`. It
+      clears the generalized hypotheses.
 
-.. tacv:: generalize {+, @term at {+ @natural} as @ident}
+   .. tacn:: dependent generalize_eqs @ident
+      :undocumented:
 
-   This is the most general form of :n:`generalize` that combines the previous
-   behaviors.
+   .. tacn:: dependent generalize_eqs_vars @ident
+      :undocumented:
 
-.. tacv:: generalize dependent @term
+   .. tacn:: generalize_eqs @ident
+      :undocumented:
 
-   This generalizes term but also *all* hypotheses that depend on :n:`@term`. It
-   clears the generalized hypotheses.
+   .. tacn:: generalize_eqs_vars @ident
+      :undocumented:
 
-.. tacn:: evar (@ident : @term)
-   :name: evar
+.. tacn:: evar ( @ident : @term )
```

Comment:
The second form is a `one_type`?  I can define that as a new nonterminal if we're going to use it in more than 1 or 2 places.

Reply:
OTOH, why did this and the others use `one_term` instead of just `term`?  For `assert`, `by` and `as` are both keywords.  Maybe we should update the mlgs instead?

Reply:
It's true that all these tactics could just as well accept a `term` / `type` without the surrounding parentheses. But I guess there is a bit of tradition in the Coq tactics that they usually take their arguments as `one_term` when they are not already in a context where they are surrounded by parentheses... Feel free to open an issue to discuss this and to keep `one_term` in the meantime (let's just fix the issue I reported about `type` being still used in the description of `elimtype` in that case).